imageio/plugins/pillow.py

```diff
@@ -4,863 +4,191 @@
-                self._length = self._im.n_frames
-
-        def _get_file(self):
-            self._we_own_fp = False
-            return self.request.get_file()
-
-        def _close(self):
-            save_pillow_close(self._im)
-            if self._we_own_fp:
-                self._fp.close()
-            # else: request object handles closing the _fp
-
-        def _get_length(self):
-            return self._length
-
-        def _seek(self, index):
-            try:
-                self._im.seek(index)
-            except EOFError:
-                raise IndexError("Could not seek to index %i" % index)
-
-        def _get_data(self, index):
-            if index >= self._length:
-                raise IndexError("Image index %i > %i" % (index, self._length))
-            i = self._im.tell()
-            if i > index:
-                self._seek(index)  # just try
-            else:
-                while i < index:  # some formats need to be read in sequence
-                    i += 1
-                    self._seek(i)
-            if self._im.palette and self._im.palette.dirty:
-                self._im.palette.rawmode_saved = self._im.palette.rawmode
-            self._im.getdata()[0]
-            im = pil_get_frame(self._im, **self._kwargs)
-            return im, self._im.info
-
-        def _get_meta_data(self, index):
-            if not (index is None or index == 0):
-                raise IndexError()
-            return self._im.info
-
-    class Writer(Format.Writer):
-        def _open(self):
-            Image = self.format._init_pillow()
-            try:
-                self._save_func = Image.SAVE[self.format.plugin_id]
-            except KeyError:
-                raise RuntimeError("Format %s cannot write images." % self.format.name)
-            self._fp = self.request.get_file()
-            self._meta = {}
-            self._written = False
-
-        def _close(self):
-            pass  # request object handled closing _fp
-
-        def _append_data(self, im, meta):
-            if self._written:
-                raise RuntimeError(
-                    "Format %s only supports single images." % self.format.name
-                )
-            # Pop unit dimension for grayscale images
-            if im.ndim == 3 and im.shape[-1] == 1:
-                im = im[:, :, 0]
-            self._written = True
-            self._meta.update(meta)
-            img = ndarray_to_pil(
-                im, self.format.plugin_id, self._meta.pop("prefer_uint8", True)
-            )
-            if "bits" in self._meta:
-                img = img.quantize()  # Make it a P image, so bits arg is used
-            img.save(self._fp, format=self.format.plugin_id, **self._meta)
-            save_pillow_close(img)
-
-        def set_meta_data(self, meta):
-            self._meta.update(meta)
-
-
-class PNGFormat(PillowFormat):
-    """A PNG format based on Pillow.
-
-    This format supports grayscale, RGB and RGBA images.
```

Comment:
A lot of code is removed here, which will need to be revived, I think, because these are plugins for specific formats such as png and jpeg, which accept a specific set of arguments. Some of the other code is logic to deal with several specific cases, like dealing with paletted images and int16.

Reply:
Yes. I've gone through the code and most of the special cases we are handling appear to be legacy code. The recent version of pillow takes care of most things just fine or fixed the bug(s) that made us introduce a workaround. 

There was one thing that pillow doesn't offer (or at least I couldn't find a matching feature) and that is `subrectangles` for GIF. I started implementing it, but then I realized that this is a completely dedicated GIF writer. Correspondingly, I think it should be its own plugin. The pillow plugin does whatever pillow is capable of, and wherever we go beyond that, we communicate that it's not pillow doing the magix, so users don't get surprised. We can make this a separate PR (preferred) or I can add it here. So far I left comments where I messed with the corresponding tests.

Speaking of tests, I faithfully ported all of the old ones for the new plugin, so there shouldn't be a regression in feature-completeness assuming that the tests were complete in that regard.

Reply:
Maybe some of the code were workarounds for bugs in Pillow, but I think a lot it was not :)

I agree that the Gif logic might deserve its own plugin (in its own module).

> [...] assuming that the tests were complete in that regard.

Some of that code is still from when imageio spun out of skimage. I bet there are use-cases covered with the code that are not actually covered by tests ...

Reply:
> I bet there are use-cases covered with the code that are not actually covered by tests ...

Fan. I will go through the legacy `pillow.py` and `pillowmulti.py` and check if I can spot any functionality we are not covering by tests yet. 

That said, The old `pillow` plugin is still available via `imageio.imread/mimread/.../get_reader` or via `imopen(uri, format="PIL-PNG")` etc.

Perhaps I will go through those files after I fixed the coverage reports. This way, I can at least see which lines aren't covered by tests yet and (correspondingly) find where hidden features might be lurking.